### ISSUE TITLE
Independent critic update issue

### ISSUE THREAD

**Issue by HshineO** [OpenRobotLab/HoST#5]
Hi ! Thanks for your excellent work and open-source code ! But it seems that I find a bug about independent critic update in multi-critics.

In the function `update` of `ppo.py` , the parts of Value function loss compute is 
```python
value_loss = (returns_batch - value_batch).pow(2).mean()
```

and the `value_batch` is computed by 
```python
value = torch.concat([critic(critic_observations) for critic in self.multi_critics], dim=-1)
```
in function `evaluate` of `actor_critics.py`.

You concatenated the outputs of the multiple critic networks into a single `value_batch` of shape [:, num_critics], then computed a global mean squared error `value_loss`  and called `backward()`.  This causes all critics to receive gradients from the other critic branches as well, meaning the updates to these critics are coupled, which violates the design principle where each critic is supposed to independently predict and be updated based on its own reward group signal.

I have also confirmed this issue through experiments in my own project.

**Comment by TaoHuang13** [comment]
Hi! Thanks for raising this point, it's appreciated that you're exploring the multi-critic setup deeply. 

We think the gradients are not coupled. `value_batch` is of shape `[batch_size, num_crtics]`, and same shape `[batch_size, num_crtics]` for `return_batch`.

In the current implementation, each critic maintains a separate computation graph, and during backpropagation, PyTorch will only pass gradients to the parameters that contributed to each element in value. So, although we compute the value loss in a batched way, each critic is still updated only based on its own prediction and corresponding return. Then the gradients remain independent, and no unwanted coupling occurs.

Please tell us if you have different opinions and results!

**Comment by HshineO** [comment]
Thank you for your patient reply!  But I still do not understand why each critic maintains a separate computation graph, I think that because the loss is computed globally by aggregating all critics' outputs into a single mean squared error, PyTorch will construct a joint computation graph that mixes all outputs. During the backward pass, although each critic ideally only affects its own output, in practice, the autograd engine tracks all operations collectively, which can introduce unwanted dependencies, subtle gradient contamination, or inefficient computation graphs.

I simply modified the calculation method of value loss from
```python
# Value function loss
                if self.use_clipped_value_loss:
                    value_clipped = target_values_batch + (value_batch - target_values_batch).clamp(-self.clip_param,
                                                                                                    self.clip_param)
                    value_losses = (value_batch - returns_batch).pow(2)
                    value_losses_clipped = (value_clipped - returns_batch).pow(2)
                    value_loss = torch.max(value_losses, value_losses_clipped).mean()
                else:
                    value_loss = (returns_batch - value_batch).pow(2).mean()
```
to
```python
                value_loss_critics = []
                for i in range(self.actor_critic.num_critics):
                    if self.use_clipped_value_loss:
                        value_clipped = target_values_batch[:, i] + (value_batch[:, i] - target_values_batch[:, i]).clamp(
                            -self.clip_param, self.clip_param
                        )
                        value_losses = (value_batch[:, i] - returns_batch[:, i]).pow(2)
                        value_losses_clipped = (value_clipped - returns_batch[:, i]).pow(2)
                        value_loss_critics.append(torch.max(value_losses, value_losses_clipped).mean())
                    else:
                        value_loss_critics.append((returns_batch[:, i] - value_batch[:, i]).pow(2).mean())
                
                value_loss = sum(value_loss_critics)
```
This ensures that each critic receives gradients only from its own output and the computation graph stays clean and independent.

I conducted a comparative experiment using your open - source code. The reward curve of the modified version has increased significantly.

<img src="https://github.com/user-attachments/assets/d9f80b32-bd8c-410c-bc02-e7044c488a19" width="50%">

So I think that compute value loss in a global batched mean square way will cause  the updates to multiple critics be coupled.

Looking forward to your further reply!

**Comment by why618188** [comment]
Hi, let me elaborate on the question in more detail.

In our implementation, the `value_batch` is computed as follows:
```python
value = torch.concat([critic(critic_observations) for critic in self.multi_critics], dim=-1)
```
and the value loss used for training is:
```python
value_loss = (returns_batch - value_batch).pow(2).mean().
```
To illustrate this in mathematical form, let’s consider the case with two critics. The total loss can be written as: $$L=\frac{1}{2N}\sum\limits_{i=1}^{N}\left[(v_{i,1}-r_{i,1})^{2}+(v_{i,2}-r_{i,2})^{2}\right]=\frac{1}{2}(L_{1}+L_{2}),$$
where $i$ indexes over the samples in the batch, $N$ is the batch size, $v_{i,1}$ is the output of the first critic, $r_{i,1}$ is the corresponding target value from `return batch`, and $L_1=\frac{1}{N}\sum\limits_{i=1}^{N}(v_{i,1}-r_{i,1})^{2}$ is the MSE loss for the first critic.

The gradient of this loss with respect to the parameters $\theta_{1}$ of the first critic is $\nabla_{\theta_1}L=\frac{1}{2}\nabla_{\theta_1}L_{1}$. This indicates that the gradient for the first critic depends only on its own loss term $L_1$, and not on the loss from the second critic. **The .mean() operation does not introduce gradient coupling, as it merely performs element-wise summation followed by a scalar division.**

For your implementation,
```python
value_loss_critics = []
for i in range(self.actor_critic.num_critics):
	value_loss_critics.append((returns_batch[:, i] - value_batch[:, i]).pow(2).mean())
	         
value_loss = sum(value_loss_critics)
```
Mathematically, this corresponds to $L^{\prime}=L_{1}+L_{2}$. Then, the gradient with respect to the first critic is $\nabla_{\theta_{1}}L^{\prime}=\nabla_{\theta_{1}}L_{1}$, which also has no gradient coupling.
The only difference between $L$ and $L^{\prime}$ is a constant scaling factor, i.e., $\nabla_{\theta_1}L=2\nabla_{\theta_{1}}L^\prime$. This scalar difference arises from the use of `.mean()` vs. `.sum()`, and does not affect the direction of the gradient, only its magnitude.

So why the performance is different? We guess this is due to the effect of the scaling factor on the gradients, which plays a role similar to that of a `reward_group_weight`. In practice, this scalar may act as an implicit weighting term, which can influence the learning dynamics—especially in reward learning or multi-task settings. Therefore, in your current setup, the latter formulation may lead to more stable or appropriate training behavior.

**Comment by HshineO** [comment]
@why618188 Thank you for your patient reply! I believe that I now fully understand the question!
